### DISCUSSION TITLE
How can I make my bot work with user-installed commands?

### DISCUSSION THREAD

**Discussion by heysailboat** [Rapptz/discord.py#9848] [Q&A] [answered]
Added my bot to my user but the commands aren't showing up. How do I do that?

**Comment by MCausc78** [comment, marked as answer]
By [syncing the tree](https://discordpy.readthedocs.io/en/latest/interactions/api.html#discord.app_commands.CommandTree.sync) and using [`app_commands.user_install()`](https://discordpy.readthedocs.io/en/latest/interactions/api.html#discord.app_commands.user_install) decorator

**Reply by heysailboat** [reply]
app_command has no attribute user_install

**Reply by MCausc78** [reply]
you need to install master branch:
```shell
$ pip install git+https://github.com/Rapptz/discord.py@master
```

**Reply by tjf1dev** [reply]
i mean its definitely not working

**Reply by pythonmcpi** [reply]
User-installable commands are working fine for me when using the latest commit on master.

**Reply by MatrixCodeDE** [reply]
> i mean its definitely not working

You have to specifically install it with the command @MCausc78 mentioned. If you try to install it with `pip install discord.py` you won't get the latest version (2.4). The installation with the git clone will get you the latest version.

**Reply by JTMC-G3** [reply]
> > i mean its definitely not working
> 
> You have to specifically install it with the command @MCausc78 mentioned. If you try to install it with `pip install discord.py` you won't get the latest version (2.4). The installation with the git clone will get you the latest version.

Does not work for me. can you send a code example?

**Reply by RequiemB** [reply]
Install the master branch using this:
```bash
pip install -U git+https://github.com/Rapptz/discord.py@master
```

Then, test it with this code
```python
@app_commands.command()
@app_commands.allowed_installs(guilds=False, users=True)
@app_commands.allowed_contexts(guilds=True, dms=True, private_channels=True)
async def useable_only_users(interaction: discord.Interaction):
    await interaction.response.send_message("I am only installed to users, but can be used anywhere.")
```

**Reply by JTMC-G3** [reply]
> Install the master branch using this:
> 
> ```shell
> pip install -U git+https://github.com/Rapptz/discord.py@master
> ```
> 
> Then, test it with this code
> 
> ```python
> @app_commands.command()
> @app_commands.allowed_installs(guilds=False, users=True)
> @app_commands.allowed_contexts(guilds=True, dms=True, private_channels=True)
> async def useable_only_users(interaction: discord.Interaction):
>     await interaction.response.send_message("I am only installed to users, but can be used anywhere.")
> ```
I did exactly what you did. The code did not work. Can i get like a full main.py that works for you?

Here's mine
(fair warning i used a modal template)

```
import discord
from discord import app_commands

import traceback

# The guild in which this slash command will be registered.
# It is recommended to have a test guild to separate from your "production" bot
TEST_GUILD = discord.Object(0)


class MyClient(discord.Client):
    def __init__(self) -> None:
        # Just default intents and a `discord.Client` instance
        # We don't need a `commands.Bot` instance because we are not
        # creating text-based commands.
        intents = discord.Intents.default()
        super().__init__(intents=intents)

        # We need an `discord.app_commands.CommandTree` instance
        # to register application commands (slash commands in this case)
        self.tree = app_commands.CommandTree(self)

    async def on_ready(self):
        print(f'Logged in as {self.user} (ID: {self.user.id})')
        print('------')

    async def setup_hook(self) -> None:
        # Sync the application command with Discord.
        await self.tree.sync()


class Feedback(discord.ui.Modal, title='Feedback'):
    # Our modal classes MUST subclass `discord.ui.Modal`,
    # but the title can be whatever you want.

    # This will be a short input, where the user can enter their name
    # It will also have a placeholder, as denoted by the `placeholder` kwarg.
    # By default, it is required and is a short-style input which is exactly
    # what we want.
    name = discord.ui.TextInput(
        label='Name',
        placeholder='Your name here...',
    )

    # This is a longer, paragraph style input, where user can submit feedback
    # Unlike the name, it is not required. If filled out, however, it will
    # only accept a maximum of 300 characters, as denoted by the
    # `max_length=300` kwarg.
    feedback = discord.ui.TextInput(
        label='What do you think of this new feature?',
        style=discord.TextStyle.long,
        placeholder='Type your feedback here...',
        required=False,
        max_length=300,
    )

    async def on_submit(self, interaction: discord.Interaction):
        await interaction.response.send_message(f'Thanks for your feedback, {self.name.value}!', ephemeral=True)

    async def on_error(self, interaction: discord.Interaction, error: Exception) -> None:
        await interaction.response.send_message('Oops! Something went wrong.', ephemeral=True)

        # Make sure we know what the error actually is
        traceback.print_exception(type(error), error, error.__traceback__)


client = MyClient()


@client.tree.command(name="submitfeedback", description="Submit feedback")
async def feedback(interaction: discord.Interaction):
    # Send the modal with an instance of our `Feedback` class
    # Since modals require an interaction, they cannot be done as a response to a text command.
    # They can only be done as a response to either an application command or a button press.
    await interaction.response.send_modal(Feedback())


@app_commands.command()
@app_commands.allowed_installs(guilds=False, users=True)
@app_commands.allowed_contexts(guilds=True, dms=True, private_channels=True)
async def useable_only_users(interaction: discord.Interaction):
    await interaction.response.send_message("I am only installed to users, but can be used anywhere.")

@client.tree.command()
@app_commands.user_install()
async def my_user_install_command(interaction: discord.Interaction) -> None:
    await interaction.response.send_message('I am installed in users by default!')


client.run('token')
```

**Reply by MCausc78** [reply]
> > Install the master branch using this:
> > ```shell
> > pip install -U git+https://github.com/Rapptz/discord.py@master
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then, test it with this code
> > ```python
> > @app_commands.command()
> > @app_commands.allowed_installs(guilds=False, users=True)
> > @app_commands.allowed_contexts(guilds=True, dms=True, private_channels=True)
> > async def useable_only_users(interaction: discord.Interaction):
> >     await interaction.response.send_message("I am only installed to users, but can be used anywhere.")
> > ```
> 
> I did exactly what you did. The code did not work. Can i get like a full main.py that works for you?
> 
> Here's mine (fair warning i used a modal template)
> 
> ```
> import discord
> from discord import app_commands
> 
> import traceback
> 
> # The guild in which this slash command will be registered.
> # It is recommended to have a test guild to separate from your "production" bot
> TEST_GUILD = discord.Object(0)
> 
> 
> class MyClient(discord.Client):
>     def __init__(self) -> None:
>         # Just default intents and a `discord.Client` instance
>         # We don't need a `commands.Bot` instance because we are not
>         # creating text-based commands.
>         intents = discord.Intents.default()
>         super().__init__(intents=intents)
> 
>         # We need an `discord.app_commands.CommandTree` instance
>         # to register application commands (slash commands in this case)
>         self.tree = app_commands.CommandTree(self)
> 
>     async def on_ready(self):
>         print(f'Logged in as {self.user} (ID: {self.user.id})')
>         print('------')
> 
>     async def setup_hook(self) -> None:
>         # Sync the application command with Discord.
>         await self.tree.sync()
> 
> 
> class Feedback(discord.ui.Modal, title='Feedback'):
>     # Our modal classes MUST subclass `discord.ui.Modal`,
>     # but the title can be whatever you want.
> 
>     # This will be a short input, where the user can enter their name
>     # It will also have a placeholder, as denoted by the `placeholder` kwarg.
>     # By default, it is required and is a short-style input which is exactly
>     # what we want.
>     name = discord.ui.TextInput(
>         label='Name',
>         placeholder='Your name here...',
>     )
> 
>     # This is a longer, paragraph style input, where user can submit feedback
>     # Unlike the name, it is not required. If filled out, however, it will
>     # only accept a maximum of 300 characters, as denoted by the
>     # `max_length=300` kwarg.
>     feedback = discord.ui.TextInput(
>         label='What do you think of this new feature?',
>         style=discord.TextStyle.long,
>         placeholder='Type your feedback here...',
>         required=False,
>         max_length=300,
>     )
> 
>     async def on_submit(self, interaction: discord.Interaction):
>         await interaction.response.send_message(f'Thanks for your feedback, {self.name.value}!', ephemeral=True)
> 
>     async def on_error(self, interaction: discord.Interaction, error: Exception) -> None:
>         await interaction.response.send_message('Oops! Something went wrong.', ephemeral=True)
> 
>         # Make sure we know what the error actually is
>         traceback.print_exception(type(error), error, error.__traceback__)
> 
> 
> client = MyClient()
> 
> 
> @client.tree.command(name="submitfeedback", description="Submit feedback")
> async def feedback(interaction: discord.Interaction):
>     # Send the modal with an instance of our `Feedback` class
>     # Since modals require an interaction, they cannot be done as a response to a text command.
>     # They can only be done as a response to either an application command or a button press.
>     await interaction.response.send_modal(Feedback())
> 
> 
> @app_commands.command()
> @app_commands.allowed_installs(guilds=False, users=True)
> @app_commands.allowed_contexts(guilds=True, dms=True, private_channels=True)
> async def useable_only_users(interaction: discord.Interaction):
>     await interaction.response.send_message("I am only installed to users, but can be used anywhere.")
> 
> @client.tree.command()
> @app_commands.user_install()
> async def my_user_install_command(interaction: discord.Interaction) -> None:
>     await interaction.response.send_message('I am installed in users by default!')
> 
> 
> client.run('token')
> ```

please provide more information than saying "not working". If you want real-time support, you can join discord.py server: https://discord.gg/dpy and ask question in `#help` forum

**Reply by JTMC-G3** [reply]
More information: Authorizing the bot works. But going into a gc and typing slash appears nothing for my bot only esmBot that i authorized

**Reply by MCausc78** [reply]
Restart your discord client

**Reply by JTMC-G3** [reply]
ive tried that. Can i just get a main.py example?

**Reply by RequiemB** [reply]
It's not working because `@app_commands.command()` does not add the command to the tree, either do `client.tree.add_command(your_command)` or replace the `@app_commands.command()` decorator with `@client.tree.command()`.

**Reply by JTMC-G3** [reply]
For some reason. Literally everything works now. It would not work for like 26 days from that reply but it works now and it's pretty darn cool

**Reply by heysailboat** [reply]
> you need to install master branch:
> 
> ```shell
> $ pip install git+https://github.com/Rapptz/discord.py@master
> ```

perfect. works like a charm.

**Comment by potatobyme** [comment]
Anyone can provide a example code

**Reply by Soheab** [reply]
[This](https://about.abstractumbra.dev/discord.py/2024/04/11/user-installable-applications.html) blog post may help.

Side note: Please do not necrobump old, answered discussions next time.  Consider joining the [server](https://discord.gg/dpy) for any questions.
